plugin/builtin/patchstrategicmergetransformer/PatchStrategicMergeTransformer_test.go

```diff
@@ -302,15 +302,15 @@ spec:
         old-label: old-value
     spec:
       containers:
+      - image: busybox
+        name: busybox
```

Comment:
Thank you for the PR! I'm a bit puzzled by some of the changes. Maybe I am missing something but it seems as though the order of containers has changed from the input patch to the output here?

Reply:
Thanks for taking a look, @natasha41575

The order of the containers is changed because of the behavior of `MergeOptionsListPrepend` and how the merge operation considers the first resource in the slice to be the original and subsequent resources to be patches.

In this case, the first file called `patch1.yaml` (confusingly named) is considered the original..  It defines:
```
spec:
  containers:
  - name: nginx
```

The second file, `patch2.yaml` defines:
```
spec:
  containers:
  - name: nginx
  - name: busybox
```

Performing the merge:
1. First iteration processes `patch1.yaml` and sees the `nginx` container.
2. Second iteration processes `patch2.yaml`.  It sees that `nginx` is already there so nothing needs to be done.  Then it sees `busybox` is not there yet and prepends it.

The result is:
```
spec:
  containers:
  - name: busybox
  - name: nginx
```

I can see why it is surprising.  

It is preserving the order of the original and only prepending the new, unseen values, and if the patch has some existing values and some new values, the result ends up looking like this.

My intention was to solve the problem where the original order of values is not preserved when merging, but this scenario seems to be a side effect of doing that.

Do you think this is the correct behavior when a patch defines a list where some elements match the original and others are new?

Reply:
Tricky issue! I read over the details, and I disagree that the original reported behavior is a bug: if an SMP lists the containers in a different order than the base, IMO they should be reordered. However, further down the issue a commenter mentions that the order changes if you have multiple containers and the SMP targets just one of them. That would definitely be a bug. I think there are two possible sources of precedent here: the older version of Kustomize that reportedly behaved differently, and Kubectl's patching behavior. I'm inclined to say the latter is canonical, so I tried creating a deployment and seeing what `kubectl patch` (v1.21) does in those two scenarios. Indeed, it allows the SMP to reorder the containers, and it does not reorder when a subset of containers is sent.

Reply:
@KnVerey Thanks for taking a look.  I think I understand what you're saying, but just to be sure let me attempt to restate and it with some examples and see if you agree:

---

# Scenario 1: Same items in patch, but in a different order

### Base:
```
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  initContainers:
    - name: foo
    - name: bar
```
### Patch:
```
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  initContainers:
    - name: bar
    - name: foo
```
### Output:
```
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  initContainers:
  - name: bar
  - name: foo
```
### Conclusion:
This is working as expected because the patch is explcitly reordering the initContainers, so we are considering that to be a request to reorder them.

---
# Scenario 2: Subset of items in patch

### Base:
```
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  initContainers:
    - name: foo
    - name: bar
```
### Patch:
```
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  initContainers:
    - name: bar
```
### Output:
```
apiVersion: v1
kind: Pod
metadata:
  name: test
spec:
  initContainers:
  - name: bar
  - name: foo
```
### Conclusion:
This is a bug. It should retain the original order the initContainers specified in the base.

Reply:
Out of curiosity I went back through some of the older kustomize versions.  Kustomize version 2.1.0 produces what looks to be the correct output in both of the above scenarios.  I don't know if that is the last version that did, but it is a change in behavior at some point.

Reply:
Yes, your scenarios reflect what I meant. I expect this problem was introduced with the change to kyaml, which was opt-in in late 3.X versions and enabled in 4.0.

Reply:
Ok, thanks.

I will revisit this PR with the scenarios in mind and see if I can come up with something.